Boundary gauge transform

moxcodes opened the pull request:
## Proposed changes

Adds the gauge transformation utilities to perform regularity-preserving Cauchy Characteristic Extraction.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies
- [x] #1869

moxcodes committed:
Add regularity-preserving gauge transforms

moxcodes committed:
Add make_const_view for vector types

moxcodes committed:
Add make_const_view for SpinWeighted

moxcodes committed:
Add prefix tag for gauge-transformed boundary data

moxcodes committed:
Add tag for SwshInterpolator

moxcodes committed:
clang-tidy SWSH utlities

kidder commented on src/NumericalAlgorithms/Spectral/SwshTags.hpp:
```diff
@@ -192,6 +196,24 @@ struct NumberOfRadialPoints : db::SimpleTag {
+template <typename Tag>
+struct SwshInterpolator : db::SimpleTag, db::PrefixTag {
+  using tag = Tag;
+  using type = SwshInterpolator;
```
test the tag

kidder commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -439,12 +439,29 @@ operator/(const get_vector_element_type_t<T>& lhs,
   return {conj(value.data())};
 }
 
+/// Take the exponential of the spin-weighted quantity; only valid for
```
split this into separate commit with its test

kidder reviewed (commented):
can you split this into multiple commits (basically by file + associated tests)?   We've reviewed all bit this commit

moxcodes committed:
Add exp, sqrt for spin-weight 0

moxcodes committed:
move boundary test utility to shared test helpers

moxcodes committed:
Minor SwshInterpolator tweaks

- add copy assignment/constructors
- const the 'morally const' `interpolate` function

moxcodes committed:
Add gauge-related CCE tags

moxcodes commented:
Commits split/reorganized. The commits that have changed or are new:
- `Add tag for SwshInterpolator` - put in a change that was in a later commit erroneously
- `Add exp, sqrt for spin-weight 0`
- `move boundary test utility to shared test helpers`
- `Minor SwshInterpolator tweaks`
- `Add gauge-related CCE tags`
- `Add regularity-preserving gauge transforms` (other commits moved out of this one)
- `fixup: Test SwshTags`

kidder commented on src/NumericalAlgorithms/Spectral/SwshInterpolation.cpp:
```diff
@@ -417,10 +417,15 @@ void SwshInterpolator::interpolate(
+  // allow it to be 'const' anyways is that no interface makes any assumption
+  // about the starting state of the internal buffer; it is kept exclusively to
+  // save allocations.
+  libsharp_modes.set_data_ref(const_cast<std::complex<double>*>(  // NOLINT
```
maybe just make the buffer mutable?

kidder commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -114,6 +121,36 @@ struct Du : db::PrefixTag, db::SimpleTag {
+/// The conformal factor associated with an angular transformation
+struct GaugeOmega : db::SimpleTag {
+  using type = Scalar<SpinWeighted<ComplexDataVector, 0>>;
+  static std::string name() noexcept { return "GaugeOmega"; }
```
remove the name functions that are just the name of the struct

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,815 @@
+
+#pragma once
+
+#include "DataStructures/DataBox/Prefixes.hpp"
```
are these all the includes you need?

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,815 @@
+using gauge_adjustments_setup_tags =
+    tmpl::list<Tags::BondiR, Tags::BondiJ, Tags::Dr<Tags::BondiJ>>;
+
+template <typename Tag>
```
cond the declaration;  should these go in a doxygen group?

moxcodes replied:
I could make a Cce group, there isn't one currently; I've been using the namespace to collect things together.

nilsdeppe replied:
Up to :) We just wanted to check

nilsdeppe replied:
ping on cond the declaration

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,815 @@
+/*!
+ * \brief Computes the evolution gauge quantity \f$\hat J\f$ on the worldtube
+ *
+ * \details The evolution gauge quantity \f$\hat J\f$ obets
```
obeys

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,815 @@
+      const gsl::not_null<tnsr::i<DataVector, 3>*>
+          cartesian_coordinates) noexcept {
+    // normalize the cartesian coordinates
+    DataVector one_over_cartesian_r =
```
const

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,815 @@
+ * `GaugeFactorSpin2` is the gauge factor \f$\hat c\f$ and takes the value
+ *
+ * \f{align*}{
+ * \hat c = \hat q^{\hat A} \partial_{\hat A} x^A q_A,
```
it is not clear if `q_A` is being differentiated in either of the two equations

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,815 @@
+        tmpl::list<::Tags::TempiJ<0, 3>,
+                   ::Tags::TempiJ<0, 2, ::Frame::Spherical<::Frame::Inertial>>>>
+        tensor_buffers{number_of_angular_points};
+    auto angular_derivative_cartesian_source_coordinates =
```
`auto&` these two

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,815 @@
+  using argument_tags =
+      tmpl::list<CartesianCoordinateTag, Spectral::Swsh::Tags::LMax>;
+
+  static void apply(
```
please move the implementation to a cpp file (by declaring an _impl function in a detail namespace)

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,552 @@
+  // Allocation
+  SpinWeighted<ComplexDataVector, 0> r_buffer =
+      get(evolution_gauge_r) / get(omega);
+  auto eth_r = Spectral::Swsh::angular_derivative<Spectral::Swsh::Tags::Eth>(
```
const

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,552 @@
+      get(cauchy_gauge_du_r_divided_by_r));
+
+  // Allocation
+  SpinWeighted<ComplexDataVector, 0> r_buffer =
```
const

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,815 @@
+ * \f{align*}{
+ * \hat J = \frac{1}{4 \hat{\omega}^2} \left( \bar{\hat d}^2  J(\hat x^{\hat A})
+ * + \hat c^2 \bar J(\hat x^{\hat A})
+ * + 2 \hat c \bar{\hat d} K(\hat x^{\hat A}) \right)
```
should K be J?

moxcodes replied:
nope, but I'll add the clarification here that K = sqrt(1 + J \bar J).

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,552 @@
+  evolution_gauge_k.data() =
+      sqrt(1.0 + evolution_gauge_j.data() * conj(evolution_gauge_j.data()));
+
+  *evolution_gauge_q =
```
we cannot connect the documentation to the math here ...

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,552 @@
+  make_const_view(make_not_null(&evolution_gauge_j), get(volume_j), 0,
+                  Spectral::Swsh::number_of_swsh_collocation_points(l_max));
+
+  get(*evolution_gauge_u) =
```
again the math and documentation don't match...

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,552 @@
+  make_const_view(make_not_null(&evolution_gauge_j), volume_j, 0,
+                  Spectral::Swsh::number_of_swsh_collocation_points(l_max));
+
+  *evolution_gauge_w +=
```
again math != doc

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,552 @@
+      2.0 * evolution_gauge_du_r_divided_by_r * (evolution_gauge_dy_j);
+}
+
+void GaugeUpdateTimeDerivatives::apply(
```
we have no clue halp

moxcodes replied:
hmmm. Yeah, I see how the documentation is a little sparse for verifying what's going on here :P

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,552 @@
+               sqrt(1.0 + evolution_gauge_j * conj(evolution_gauge_j)));
+}
+
+void GaugeAdjustedBoundaryValue<Tags::BondiH>::apply_impl(
```
math != doc

kidder commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,815 @@
+ * identity transform of angular coordinates. Using this initialization function
+ * ensures that the evolution gauge and the Cauchy gauge angular coordinates
+ * agree on the first evaluated time.
+ */
```
document what this means in terms of numbers

kidder commented on tests/Unit/Evolution/Systems/Cce/Test_GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,538 @@
+        const auto& collocation = Spectral::Swsh::cached_collocation_metadata<
+            Spectral::Swsh::ComplexRepresentation::Interleaved>(l_max);
+        for (const auto& collocation_point : collocation) {
+          auto rootfind = boost::math::tools::bisect(
```
why not use the toms rootfind that we have wrapped

kidder commented on tests/Unit/Evolution/Systems/Cce/Test_GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,538 @@
+      {value_dist(*gen), value_dist(*gen), value_dist(*gen)}};
+  const std::array<double, 3> center{
+      {value_dist(*gen), value_dist(*gen), value_dist(*gen)}};
+  gr::Solutions::KerrSchild solution{mass, spin, center};
```
const

kidder commented on tests/Unit/Evolution/Systems/Cce/Test_GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,538 @@
+        make_not_null(&forward_and_inverse_interpolated_omega_cd),
+        get(interpolated_forward_omega_cd));
+
+    auto check_rhs =
```
const auto&

kidder commented on tests/Unit/Evolution/Systems/Cce/Test_GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,538 @@
+        interpolator.interpolate(make_not_null(&evolution_coords_u),
+                                 get(*boundary_u));
+
+        ComplexDataVector minus_u =
```
const

kidder commented on tests/Unit/Evolution/Systems/Cce/Test_GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,538 @@
+      ](const gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, 1>>*>
+            bondi_u,
+        const Scalar<SpinWeighted<ComplexDataVector, 1>>& boundary_u) noexcept {
+        ComplexDataVector time_transform = 10.0 * get(boundary_u).data();
```
const

kidder commented on tests/Unit/Evolution/Systems/Cce/Test_GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,538 @@
+           0.5 * (get(forward_u_0) * conj(get(forward_eth_omega_cd)) +
+                  conj(get(forward_u_0)) * get(forward_eth_omega_cd))) /
+          forward_omega_cd;
+  Spectral::Swsh::SwshInterpolator omega_interpolator{
```
const

kidder commented on tests/Unit/Evolution/Systems/Cce/Test_GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,538 @@
+      number_of_angular_grid_points};
+  omega_interpolator.interpolate(make_not_null(&omega_comparison_lhs),
+                                 forward_adjusted_du_omega_cd_over_omega);
+  SpinWeighted<ComplexDataVector, 0> omega_comparison_rhs =
```
const

kidder commented on src/NumericalAlgorithms/Spectral/SwshDerivatives.hpp:
```diff
@@ -451,8 +451,7 @@ template <
     typename... ArgumentTypes>
-void angular_derivatives(const size_t l_max,
-                         const size_t number_of_radial_points,
+void angular_derivatives(size_t l_max, size_t number_of_radial_points,
```
nope clang-tidy is wrong...

kidder reviewed (changes requested):
joint with @nilsdeppe

moxcodes commented:
fixups posted. I'll rebase on develop when I squash.
Cheers!

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+           gauge_c * conj(*evolution_gauge_q)) -
+      (eth_omega * evolution_gauge_k - conj(eth_omega) * evolution_gauge_j) *
+          exp(2.0 * evolution_gauge_beta) *
+          (-1.0 + (evolution_gauge_dy_j * conj(evolution_gauge_dy_j) -
```
Check this `-1.0` since it is a +1.0 in the dox

moxcodes replied:
Checked with the expressions in my paper draft, and -1.0 is correct, so I'll update the dox

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+                   0.25 *
+                       square(evolution_gauge_j * conj(evolution_gauge_dy_j) +
+                              evolution_gauge_dy_j * conj(evolution_gauge_j)) /
+                       (1.0 + evolution_gauge_j * conj(evolution_gauge_j)))) /
```
use `evolution_gauge_k`

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,759 @@
+ * \f}
+ *
+ * where the explicit argument \f$\hat x^{\hat A}\f$ on the right-hand side
+ * implies the need for an interpolation operation.
```
Can you please define `K` here too?

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,759 @@
+ * \left(\hat J \hat{\bar{\eth}} \hat \omega
+ * - \hat K \hat \eth \hat \omega\right)
+ * \left(1 + \partial_{\hat y} \hat{\bar{J}} \partial_{\hat y} \hat J
+ * - \frac{\partial_{\hat y}(\hat J \hat{\bar{J}})^2}{4 \hat K^2}\right)
```
Can you square the whole fraction to make it clear that its the deriv squared and not the deriv of the squared quantity?

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+          (conj(gauge_d) * *evolution_gauge_q -
+           gauge_c * conj(*evolution_gauge_q)) -
+      (eth_omega * evolution_gauge_k - conj(eth_omega) * evolution_gauge_j) *
+          exp(2.0 * evolution_gauge_beta) *
```
factor out e^(2beta)/(omega r^2) prefactor (division is expensive) for last two terms

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,759 @@
+ * \notag \\
+ * & + 2 \frac{e^{2 \hat \beta}}{\hat \omega \hat r^2}
+ * \left[ \hat{\bar{\eth}} \hat \omega \partial_{\hat y} \hat J
+ * + \hat{\bar{\eth}} \hat \omega \left(-\frac{\hat J \partial_{\hat y}
```
no bar on \eth in omega deriv

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,759 @@
+ * \f$\mathcal U\f$ is computed by,
+ *
+ * \f{align*}{
+ * \mathcal U = \frac{1}{2\hat \omega} \left(\hat{\bar d} U(\hat x^{\hat A})
```
check power of omega in denom on first term

moxcodes replied:
checked against derivation and \omega^2 is correct, so I'll fix the dox.

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,759 @@
+ * \mathcal U = \frac{1}{2\hat \omega} \left(\hat{\bar d} U(\hat x^{\hat A})
+ * - \hat c \bar U(\hat x^{\hat A}) \right)
+ * - \frac{e^{2 \hat \beta}}{\hat r \hat \omega}
+ * \left(\hat \eth \hat \omega \hat K
```
Can you move non-differentiated to left of derivative

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+
+  *evolution_gauge_w +=
+      (omega - 1.0) / evolution_gauge_r - 2.0 * du_omega / omega -
+      (conj(eth_omega) * (evolution_gauge_u - evolution_gauge_u_at_scri) +
```
Can you add a code comment what u - u_at_scri is and how it relates to the dox?

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,759 @@
+ * \f{align*}{
+ * \hat W =& W(\hat x^{\hat A}) + (\hat \omega - 1) / \hat r
+ * + \frac{e^{2 \hat \beta}}{2 \hat \omega^2 \hat r}
+ * \left(\hat{\bar \eth} \hat \omega^2 \hat J
```
Can you add parens to make it clear whether the deriv is squared or the thing being operated on is squared? Also can you make non-differentiated quantities be on the left of the operator for clarity?

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+             2.0 * evolution_gauge_u_at_scri * conj(eth_evolution_gauge_r) /
+                 evolution_gauge_r * evolution_gauge_dy_j +
+             evolution_gauge_u_at_scri_bar_eth_j) -
+      2.0 * cauchy_du_omega / omega * (evolution_gauge_dy_j) +
```
Can you factor this and the next line as they are in the dox?

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+  // Note also that when angular derivatives are taken numerically, they must be
+  // corrected to the Bondi \eth with jacobians proportional to \eth R / R
+  interpolator.interpolate(evolution_gauge_h, cauchy_gauge_du_j);
+  *evolution_gauge_h =
```
remove extra parens please

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,759 @@
+ * + \hat K \hat \eth \bar{\mathcal{U}}^{(0)}  \notag\\
+ * &+ \frac{1}{4 \hat \omega^2} \left(\hat{\bar d}^2 H(\hat x^{\hat A})
+ * + \hat c^2 \bar H(\hat x^{\hat A})
+ * + \hat{\bar d} c \frac{H(\hat x^{\hat A}) \bar J(\hat x^{\hat A})
```
missing `\hat` on `c`

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,759 @@
+ * \f}
+ *
+ * where the superscript \f$(0)\f$ denotes evaluation at \f$\mathcal I^+\f$ and
+ * the explicit \f$\hat x^{\hat A}\f$ arguments on the right-hand side imply
```
can you clarify what `K` and `\hat{K}` are?

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+                                   std::integral_constant<int, 0>>>(
+          computation_buffers));
+  cauchy_du_omega =
+      du_tilde_omega - 0.5 * (evolution_gauge_u_at_scri * conj(eth_omega) +
```
Get rid of `tilde` in name

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+  // using quantities of spin 2 or less.
+  // note the conversion Jacobian for the angular derivative
+  evolution_gauge_u_at_scri_bar_eth_j =
+      eth_evolution_gauge_u_at_scri_bar_j_tilde -
```
remove tilde

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,759 @@
+ * + \frac{1}{2} (\mathcal U^{(0)} \hat{\bar \eth} \hat \omega
+ * + \bar{\mathcal U}^{(0)} \hat \eth \hat \omega)
+ * \f}
+ * \note Before this update call the quantity stored in the tag
```
Let's make this `note` a warning since if one is not aware of it things will be bad.

Can you also expand on the relation between \hat{U} and U^{(0)}?

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+      get(get<::Tags::SpinWeighted<::Tags::TempScalar<2, ComplexDataVector>,
+                                   std::integral_constant<int, 0>>>(
+          computation_buffers));
+  x.data() =
```
Add comment like `// Switch to complex so we can take spin-weighted derivatives`

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.hpp:
```diff
@@ -0,0 +1,759 @@
+ * inertial angular coordinates is
+ *
+ * \f{align*}{
+ * \partial_{\hat u} x^A = \partial_{\hat A} x^{A} \mathcal U^{(0) \hat A},
```
move U^0 to left of deriv

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+      get(get<::Tags::SpinWeighted<::Tags::TempScalar<2, ComplexDataVector>,
+                                   std::integral_constant<int, 0>>>(
+          computation_buffers));
+  x.data() =
```
Please add same comment as above

nilsdeppe commented on src/Evolution/Systems/Cce/GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,685 @@
+            angular_derivative_cartesian_source_coordinates.get(i, 1);
+  }
+
+  get(*gauge_factor_spin_2).data() =
```
Add comment that these are basically `c` and `d` in the dox

nilsdeppe commented on tests/Unit/Evolution/Systems/Cce/Test_GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,538 @@
+      {value_dist(*gen), value_dist(*gen), value_dist(*gen)}};
+  const gr::Solutions::KerrSchild solution{mass, spin, center};
+
+  const double extraction_radius = 100;
```
100.0 (find others)

nilsdeppe commented on tests/Unit/Evolution/Systems/Cce/Test_GaugeTransformBoundaryData.cpp:
```diff
@@ -0,0 +1,538 @@
+  // parameters
+  const double frequency = 0.1 * value_dist(*gen);
+  const double amplitude = 0.1 * value_dist(*gen);
+  const double target_time = 50 * value_dist(*gen);
```
50.0

nilsdeppe reviewed (changes requested):
Reviewed with @kidder and @moxcodes

moxcodes commented:
rebased, old fixups squashed, and new fixup posted.
Cheers!

kidder reviewed (commented):
I am happy with the fixups

kidder commented:
but you do have clang-tidy and compiling issues...

nilsdeppe commented:
I'm also happy for you to squash

moxcodes commented:
Squashed, and added one more fixup for the CI compile fix. The clang-tidy is something in the toms 748 rootfind in the boost library, so I don't think we can do anything about that.

nilsdeppe commented:
LGTM. Yes, that's clang-tidy and boost being annoying so we just ignore that error :)

kidder commented:
yes squash

moxcodes commented:
squashed.
Cheers!

kidder commented:
ignoring boost clang-tidy issue
